File: scripts/build-shaders.py
#!/usr/bin/env python

import sys, re, os, errno

input_file = sys.argv[1]
output_file = sys.argv[2]

def mkdir_p(path):
    try:
        os.makedirs(path)
    except OSError as exc: # Python >2.5
        if exc.errno == errno.EEXIST and os.path.isdir(path):
            pass
        else: raise

shader_name, shader_type, extension = os.path.basename(input_file).split('.')

with open(input_file, "r") as f:
    data = f.read()

# Replace uniform pragmas

pragma_mapbox_regex = re.compile("(\s*)\#pragma \mapbox\: (define|initialize) (low|medium|high)p (float|vec(2|3|4)) (.*)")

def replace_uniform_pragmas(line):
    if pragma_mapbox_regex.match(line):
        params = line.split()
        if params[2] == "define":
            return """uniform {u_precision} {u_type} u_{u_name};""".format(
                    u_precision = params[3],
                    u_type = params[4],
                    u_name = params[5])
        else:
            return """    {u_precision} {u_type} {u_name} = u_{u_name};""".format(
                    u_precision = params[3],
                    u_type = params[4],
                    u_name = params[5])
    else:
        return line

data = "\n".join([replace_uniform_pragmas(line) for line in data.split("\n")])

content = """#pragma once

// NOTE: DO NOT CHANGE THIS FILE. IT IS AUTOMATICALLY GENERATED.

#include <mbgl/gl/gl.hpp>

namespace mbgl {{
namespace shaders {{
namespace {name} {{

#ifndef MBGL_SHADER_NAME_{NAME}
#define MBGL_SHADER_NAME_{NAME}
constexpr const char* name = "{name}";
#endif // MBGL_SHADER_NAME_{NAME}

#ifdef GL_ES_VERSION_2_0
constexpr const char* {type} = R"MBGL_SHADER(precision highp float;\n{data})MBGL_SHADER";
#else
constexpr const char* {type} = R"MBGL_SHADER(#version 120\n{data})MBGL_SHADER";
#endif

}} // namespace {name}
}} // namespace shaders
}} // namespace mbgl
""".format(
     name = shader_name,
     NAME = shader_name.upper(),
     type = shader_type,
     TYPE = shader_type.upper(),
     data = data,
)

mkdir_p(os.path.dirname(output_file))
with open(output_file, 'w') as f: f.write(content)

File: scripts/build-shaders.js
#!/usr/bin/env node

var path = require('path');
var fs = require('fs');
var mkdirp = require('mkdirp');

var shaderName = process.argv[2];
var inputPath = process.argv[3];
var outputPath = process.argv[4];

if (!shaderName || !inputPath || !outputPath) {
    console.warn('Not enough arguments.');
    console.warn('Usage: %s shaderName /path-to-shader-sources /output-path', path.basename(process.argv[1]));
    process.exit(1);
}

var pragmaMapboxRe = /(\s*)#pragma\s+mapbox\s*:\s+(define|initialize)\s+(low|medium|high)p\s+(float|vec(?:2|3|4))\s+(.*)/;


function applyPragmas(source) {
    return '\n' + source.split('\n').map(function(line) {
        var params = line.match(pragmaMapboxRe);
        if (params) {
            if (params[2] === 'define') {
                return params[1] + 'uniform ' + params[3] + 'p ' + params[4] + ' u_' + params[5] + ';';
            } else {
                return params[1] + params[3] + 'p ' + params[4] + ' ' + params[5] + ' = u_' + params[5] + ';';
            }
        } else {
            return line;
        }
    }).join('\n');
}

var vertexPrelude = fs.readFileSync(path.join(inputPath, '_prelude.vertex.glsl'));
var fragmentPrelude = fs.readFileSync(path.join(inputPath, '_prelude.fragment.glsl'));
var vertexSource = vertexPrelude + fs.readFileSync(path.join(inputPath, shaderName + '.vertex.glsl'), 'utf8');
var fragmentSource = fragmentPrelude + fs.readFileSync(path.join(inputPath, shaderName + '.fragment.glsl'), 'utf8');

var content = "#pragma once\n" +
"\n" +
"// NOTE: DO NOT CHANGE THIS FILE. IT IS AUTOMATICALLY GENERATED.\n" +
"\n" +
"#include <mbgl/gl/gl.hpp>\n" +
"\n" +
"namespace mbgl {\n" +
"namespace shaders {\n" +
"\n" +
"class " + shaderName + " {\n" +
"public:\n" +
"    static constexpr const char* name = \"" + shaderName + "\";\n" +
"    static constexpr const char* vertexSource = R\"MBGL_SHADER(" + applyPragmas(vertexSource) + ")MBGL_SHADER\";\n" +
"    static constexpr const char* fragmentSource = R\"MBGL_SHADER(" + applyPragmas(fragmentSource) + ")MBGL_SHADER\";\n" +
"};\n" +
"\n" +
"} // namespace shaders\n" +
"} // namespace mbgl\n";

mkdirp.sync(outputPath);
fs.writeFileSync(path.join(outputPath, shaderName + '.hpp'), content);
